### PR TITLE
feat(opts)!: restructure mappings and set default picker opts in setup

### PR DESCRIPTION
**Pull request by al-ce**
BREAKING CHANGES: In this PR we would restructure the `mappings` table in in the setup configuration to follow Telescope's structure ( `["key"] = function`) for action mappings.

All actions are named functions so they'll have names in the Telescope 'which key' popup.

Changes:

- The configuration for the Telescope picker is given some defaults that apply to the `Hbac telescope `command as well as a call to `hbac.telescope()`. Any opts added to the user's setup call or passed to the function call deep-extend into the config defaults
- Some `require` calls are moved around in `hbac.command.actions`, `hbac.command.subcommands`, and `hbac.utils` to prevent circular dependencies with the mappings in the setup configuration for the Telescope picker
- `attach_mappings` module is renamed to `actions`. It holds the definitions for the named functions that refresh the picker
- the `attach_mappings` function definition is moved to `hbac.telescope` and can accept an `opts` table so that the user can pass their own `mappings` configuration as they would to a builtin Telescope picker
- a minor fix to the `ordinal` value for the results. This ignores the absolute filepath above a project root for a cleaner fuzzy search (e.g. typing `hu` won't show `/home/user/projectroot/main.c` as a result)
- - README is updated with relevant instructions and examples

Example user config (extends default in `hbac.config`):

```lua
hbac.setup({
  telescope = {
    sort_lastused = false,  -- overrides default
    layout_strategy = "vertical",  -- additional opts
    prompt_prefix = "Hbac! ",
    mappings = {  -- extends default mappings
      i = {
        ["<CR>"] = telescope_actions.select_drop,
        ["<C-CR>"] = telescope_actions.select_default,
        ["<M-z>"] = function() print("hello") end,
      },
    },
    pin_picker = {
      -- etc.
    },
  },
})
```

example keymap for picker (extends user config):

```lua
vim.keymap.set("n", ",ht", function()
  require("hbac").telescope(
    {
      ignore_current_buffer = true,
      file_ignore_patterns = { ".json" },
      mappings = {
        i = {
          ["<CR>"] = telescope_actions.select_default,  -- overrides the mapping in user setup config
        },
      },
    }
  )
end, { desc = "Hbac: pin picker", noremap = true })
```

Addresses #11 and builds on #9 and #12 (partially addressed by f041ebabcd27aaadaebaccc9390edb56d06af731)

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -98,23 +83,77 @@ The plugin provides a [telescope.nvim](https://github.com/nvim-telescope/telesco
...
+```lua
+require("hbac").telescope({
+  file_ignore_patterns = { ".json" },
   layout_strategy = "horizontal",
-  initial_mode = "normal",
   -- etc.
 })
 ```
 
+There are some options that are not supported like `bufnr_width` and `path_display` since the results display is specialized, and we're not truly extending the standard buffer picker. Please consider contributing or opening an issue if something essential is missing.
```
Comment (axkirillov) [inline README.md]:
```suggestion
There are some options that are not supported like `bufnr_width` and `path_display`. Please consider contributing or opening an issue if something essential is missing.
```

File: lua/hbac/telescope/actions.lua [hunk 2]
```diff
@@ -21,45 +17,31 @@ local function execute_telescope_action(prompt_bufnr, action)
...
 local function hbac_delete_buffer(prompt_bufnr)
-	execute_telescope_action(prompt_bufnr, config.values.close_command)
+	local close_command = require("hbac.config").values.close_command
+	execute_telescope_action(prompt_bufnr, close_command)
 end
 
-M.attach_mappings = function(_, map)
-	local hbac_telescope_actions = {
+
+return {
```
Comment (axkirillov) [inline lua/hbac/telescope/actions.lua]: Is it important to not use M here? or why did you change it?

File: lua/hbac/telescope/init.lua [hunk 3]
```diff
@@ -16,20 +16,31 @@ local telescope_conf = require("telescope.config").values
...
-		:find()
+	pickers.new(opts, {
+		prompt_title = "Hbac Pin States",
+		finder = make_finder.make_finder(opts),
+		sorter = telescope_conf.generic_sorter(opts),
+		attach_mappings = attach_mappings(opts),
+		previewer = telescope_conf.file_previewer(opts),
+		default_selection_index = make_finder.default_selection_idx,
+	})
+			:find()
```
Comment (axkirillov) [inline lua/hbac/telescope/init.lua]: Is it alligned with the . ? Looks weird at least on github
How about
```suggestion
	}):find()
```
?

File: lua/hbac/telescope/init.lua [hunk 4]
```diff
@@ -16,20 +16,31 @@ local telescope_conf = require("telescope.config").values
...
-		.new(opts, {
-			prompt_title = "Hbac Pin States",
-			finder = make_finder.make_finder(opts),
-			sorter = telescope_conf.generic_sorter(opts),
-			attach_mappings = attach_mappings.attach_mappings,
-			previewer = telescope_conf.file_previewer(opts),
-			default_selection_index = make_finder.default_selection_idx,
-		})
-		:find()
+	pickers.new(opts, {
```
Comment (axkirillov) [inline lua/hbac/telescope/init.lua:35]: Alternatively to the other suggestion
```suggestion
	pickers
	    .new(opts, {
```

File: lua/hbac/utils.lua [hunk 5]
```diff
@@ -10,6 +7,8 @@ M.get_listed_buffers = function()
 end
 
 M.buf_autoclosable = function(bufnr)
+	local config = require("hbac.config")
+	local state = require("hbac.state")
```
Comment (axkirillov) [inline lua/hbac/utils.lua]: Is there a reason to move it here?

**Comment by axkirillov** [comment]
Looks good to me. I'll test it on my setup.

File: lua/hbac/telescope/actions.lua (hunk 2 above)
Comment (al-ce) [reply lua/hbac/telescope/actions.lua]: No we can still use it! Just oversimplifying I guess, or I was trying to imitate hbac.init.lua ? The important part is defining the named local functions and assigning them as values to the keys in the table (unless there's a better way). Just as long as we're not doing e.g.

```lua
M.pin_all = function(prompt_bufnr) -- etc. 
end

return M
```

Will change it back

File: lua/hbac/telescope/init.lua (hunk 3 above)
Comment (al-ce) [reply lua/hbac/telescope/init.lua]: Yeah that's weird, maybe something weird with my `stylua` settings, good catch. Is this ok?

```lua
	pickers.new(opts, {
		prompt_title = "Hbac Pin States",
		finder = make_finder.make_finder(opts),
		sorter = telescope_conf.generic_sorter(opts),
		attach_mappings = attach_mappings(opts),
		previewer = telescope_conf.file_previewer(opts),
		default_selection_index = make_finder.default_selection_idx,
	}):find()
```

**Comment by axkirillov** [comment]
Did you decide to make the current buffer hidden by default after all?

File: lua/hbac/utils.lua (hunk 5 above)
Comment (al-ce) [reply lua/hbac/utils.lua]: No you're right, I think this is a leftover from when I was trying to debug the require-loop, which seems to only be an issue in hbac.command.actions. I'll check but otherwise I'll put this back

File: lua/hbac/telescope/init.lua (hunk 3 above)
Comment (axkirillov) [reply lua/hbac/telescope/init.lua]: yes, this is fine

**Comment by al-ce** [comment]
> Did you decide to make the current buffer hidden by default after all?

Sorry I left that comment hanging! I think I'm on the same page with what you were saying, that we should not hide the current buffer (right?). Currently the defaults are

```lua
sort_mru = true,
sort_lastused = true,
selection_strategy = "row",
```

so the current buffer should be in the first result, selector/cursor should be on the second result which is the previously used buffer, and the rest are sorted by most recently used (if I've implemented the options correctly)

**Comment by axkirillov** [comment]
> > Did you decide to make the current buffer hidden by default after all?
> 
> Sorry I left that comment hanging! I think I'm on the same page with what you were saying, that we should not hide the current buffer (right?). Currently the defaults are
> 
> ```lua
> sort_mru = true,
> sort_lastused = true,
> selection_strategy = "row",
> ```
> 
> so the current buffer should be in the first result, selector/cursor should be on the second result which is the previously used buffer, and the rest are sorted by most recently used (if I've implemented the options correctly)

huh, weird, I think I have the current buffer hidden
can you check? my setup function is empy, so I should be getting all default behaviours

**Comment by al-ce** [comment]
> can you check? my setup function is empy, so I should be getting all default behaviours

I'm guilty of creating a million branches here, are you using the `default_opts` branch on my fork instead of the `picker_opts` branch from the last PR? (sorry I did that, I changed the merge message of the last PR and had a merge conflict so I created this new one) Otherwise I'll debug

**Comment by axkirillov** [comment]
> > can you check? my setup function is empy, so I should be getting all default behaviours
> 
> I'm guilty of creating a million branches here, are you using the `default_opts` branch on my fork instead of the `picker_opts` branch from the last PR? (sorry I did that, I changed the merge message of the last PR and had a merge conflict so I created this new one) Otherwise I'll debug

Yeah, I'm using this branch
```lua
	{
		'al-ce/hbac.nvim',
		config = function()
			require("hbac").setup()
		end,
		branch = "default_opts"
	},
```

**Comment by al-ce** [comment]
Does passing opts to the setup call or to the telescope function call change the behavior or does it throw any errors? Like

```lua
  {
    'al-ce/hbac.nvim',
    config = function()
      require("hbac").setup(
      {
        telescope = {
          sort_lastused = true,
        },
        }
    )

      vim.keymap.set("n", ",ht", function()
        require("hbac").telescope({
            sort_mru = true,
            mappings = {
              i = {
                ["<cr>"] = telescope_actions.select_drop,
              },
            },
          })
      end)

    end,
    branch = "default_opts"
  }
```

Just confirming whether it's just the default behavior or if any opts are registering

**Comment by axkirillov** [comment]
> Does passing opts to the setup call or to the telescope function call change the behavior or does it throw any errors? Like
> 
> ```lua
>   {
>     'al-ce/hbac.nvim',
>     config = function()
>       require("hbac").setup(
>       {
>         telescope = {
>           sort_lastused = true,
>         },
>         }
>     )
> 
>       vim.keymap.set("n", ",ht", function()
>         require("hbac").telescope({
>             sort_mru = true,
>             mappings = {
>               i = {
>                 ["<cr>"] = telescope_actions.select_drop,
>               },
>             },
>           })
>       end)
> 
>     end,
>     branch = "default_opts"
>   }
> ```
> 
> Just confirming whether it's just the default behavior or if any opts are registering

For me, even if I paste this configuration, the current buffer is hidden.

**Comment by al-ce** [comment]
Sorry for lack of updates, I've been traveling to visit family and I've been trying to figure this out in some downtime. I've tried a few different setups and I'm getting the behavior I'm expecting so I think it might be some mistake I made when I merged the last PR into this branch or I'm making some bad assumption about my machine. Sorry for the hold up and I'll keep trying, hope all is well

**Comment by axkirillov** [comment]
Sorry! I just realized I was passing my own opts 🤦🏻 
```lua
vim.keymap.set("n", "<leader>b", function()
    local default_opts = {
        show_all_buffers = true,
        sort_mru = true,
        sort_lastused = true,
        ignore_current_buffer = true,
        cwd_only = false,
        cwd = nil,
    }
    require("hbac").telescope(default_opts)

```

**Comment by axkirillov** [comment]
I changed it to this and it works as expected 
```lua
vim.keymap.set("n", "<leader>b", ":Hbac telescope<CR>", ns)
```

**Comment by al-ce** [comment]
> I changed it to this and it works as expected

Ok that's good! I integrated your suggestions above and made a small refactor in the [config module](https://github.com/al-ce/hbac.nvim/blob/default_opts/lua/hbac/config.lua) to not repeat the lines that assign the mappings, since they're the same for both. Is this ok?

```lua
local action_mappings = {
	["<M-c>"] = actions.close_unpinned,
	["<M-x>"] = actions.delete_buffer,
	["<M-a>"] = actions.pin_all,
	["<M-u>"] = actions.unpin_all,
	["<M-y>"] = actions.toggle_pin,
}
local defaults = {
	telescope = {
		mappings = {
			n = action_mappings,
			i = action_mappings,
                }
        -- etc
}
```

**Review by axkirillov — APPROVED** [review]

**Comment by axkirillov** [comment]
Looks good and works for me.

**Comment by al-ce** [comment]
Hey, I'm back home from my travels. I'm good with the PR as it is if you're happy with it and ready to merge it, I think we've addressed the open issues it references. Will you do the merge into main when you're ready? I didn't want to push the new mapping config structure without knowing your procedures for announcements or versioning. Thanks again for the collaboration!
